### PR TITLE
Invoicing for surplus contract days

### PR DESCRIPTION
**Pull request by akheron**

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt [hunk 1]
```diff
@@ -316,6 +341,16 @@ class NewDraftInvoiceGenerator(private val productProvider: InvoiceProductProvid
...
+        ) + surplusContractDays(
+            period,
+            child,
+            product,
+            price,
+            codes,
+            dailyFeeDivisor,
+            contractDaysPerMonth,
+            attendanceDates,
+            absences
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:353]: Are fee alterations applied to the daily price for surplus days?

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt [hunk 2]
```diff
@@ -289,19 +305,28 @@ class NewDraftInvoiceGenerator(private val productProvider: InvoiceProductProvid
...
+        val periodAttendanceDates = attendanceDates.filter { period.includes(it) }
+        if (periodAttendanceDates.isEmpty()) return listOf()
+
+        // Make sure the number of operational days in a month doesn't exceed `dailyFeeDivisor`.
+        //
+        // Example: A child has a placement to a round-the-clock unit for the first half and to a
+        // normal unit for the second half of the month. The round-the-clock unit has more operational
+        // days, so we have to make sure that we don't invoice more than `dailyFeeDivisor` days.
+        //
+        val attendanceDaysLeft = attendanceDates.take(dailyFeeDivisor).filter { period.includes(it) }.size
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:322]: Do we need `periodAttendanceDates` anymore?

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt [hunk 3]
```diff
@@ -366,45 +400,80 @@ class NewDraftInvoiceGenerator(private val productProvider: InvoiceProductProvid
...
 
         val parentLeaveAbsences = absences.filter { it.absenceType == AbsenceType.PARENTLEAVE }
         val parentLeaveDays = parentLeaveAbsences
             .filter { ChronoUnit.YEARS.between(child.dateOfBirth, it.date) < 2 }
-            .filter { period.includes(it.date) }
+            .filter { periodAttendanceDates.contains(it.date) }
 
-        return minOf(forceMajeureDays.size + parentLeaveDays.size, periodOperationalDates.size)
+        return minOf(forceMajeureDays.size + parentLeaveDays.size, periodAttendanceDates.size)
     }
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:477]: These changes currently break daily refunds for children in round the clock units. At least until planned absences are taken into account.

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt (hunk 1 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt]: Fixed

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt (hunk 2 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:322]: Not sure what you mean. They are required for computing how many "assumed attendance days" this period has of all the days in the month.

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt (hunk 3 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt]: Good catch, fixed

File: service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt (hunk 2 above)
Comment (jaahon) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:322]: I mean why don't we use
```
attendanceDates.take(dailyFeeDivisor).filter { period.includes(it) }
```
in place of 
```
attendanceDates.filter { period.includes(it) }
```

Comment (jaahon) [reply service/src/main/kotlin/fi/espoo/evaka/invoicing/service/NewDraftInvoiceGenerator.kt:322]: Maybe we should discuss this somewhere else.

**Review by jaahon — APPROVED** [review]

**Pull request merged**